File: 0238-product-of-array-except-self/0238-product-of-array-except-self.scala
object Solution {
    def productExceptSelf(nums: Array[Int]): Array[Int] = {

    var prefix = Array.fill(nums.length)(0)
    var suffix = Array.fill(nums.length)(0)
    var output = Array.fill(nums.length)(0)
    prefix(0) = 1
    suffix(nums.length-1) = 1

    for(i<-1 until nums.length){
      prefix(i) = prefix(i-1)*nums(i-1)
    }
    for(i<-nums.length-2 until -1 by -1){
      suffix(i) = suffix(i+1) * nums(i+1)
    }

    for(i<-nums.indices){
      output(i) = prefix(i) * suffix(i)
    }

    output
        
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
            
        prefix = [1]
        suffix = [1]
        output = []

        for i in range(0,len(nums)-1):
            prefix.append(prefix[-1]*nums[i])
        print(prefix)
        for i in range(len(nums)-1,0,-1):
            suffix.append(suffix[-1]*nums[i])
        print(suffix)
        for i in range(0,len(nums)):
            output.append(prefix[i]*suffix[len(nums)-1-i])
            print(prefix[i])
            print(suffix[len(nums)-1-i])

        return output
